### PR TITLE
component-initialized event

### PR DESCRIPTION
**Pull request by ngokevin**
**Description:**

Wanted a way to depend on other components dynamically. (e.g., I have a component that draws line, where the line data resides in another component. Thus the line component must wait for the other component to init, without knowing beforehand what that component is nor hardcoding it to avoid decoupling).

**Changes proposed:**
- add component-initialized event

### CODE REVIEW

**Comment by donmccurdy** [comment]
In practice, how would I use this when writing a component with these dependencies?

1. During `myComponent.init()`, check for whatever I need. 
2. If missing, listen for `component-initialized`.
3. On `component-initialized`, check for dependency again. If missing, go back to 2. Else, go to 4.
4. Finish initialization.

In cases like you the 'line data' example, consistent event names are probably most important. For example, having all loaders emit the `model-loaded` event, regardless of the format, is very helpful.

(Note – I think adding a `component-initialized` event seems worthwhile regardless. Just brainstorming whether this is the ideal way to handle dynamic component dependencies.)

**Comment by ngokevin** [comment]
I made a hacky API.

```html
<a-entity draw-line="pointer: movement-pattern.movementPattern.spline" movement-pattern">
```

It will grab the `spline` object from `el.components['movement-pattern'].movementPattern.spline` and draw it. So I needed to wait on that component to initialize, at which it'd be ready.

Maybe I should just couple them, or make `draw-line` just an internal function. This is dirty, I admit.

**Comment by donmccurdy** [comment]
Ok cool. Yeah for most cases I can think of, I'd probably use a custom event or method. But `component-initialized` seems at least as important as `componentchanged`, from a lifecycle perspective, so this event works fine.

(Difference in hyphenation is because the all-one-word thing is phasing out, i assume?)

File: src/core/component.js [hunk 1]
```diff
@@ -201,16 +201,20 @@ Component.prototype = {
       // Play the component if the entity is playing.
       this.update(oldData);
       if (el.isPlaying) { this.play(); }
+      el.emit('component-initialized', {
```
Comment (dmarcos) [inline src/core/component.js]: remove dash from the event name for consistency?

File: tests/components/raycaster.test.js [hunk 2]
```diff
@@ -80,8 +80,10 @@ suite('raycaster', function () {
       var newEl = this.targetEl = document.createElement('a-entity');
       var numObjects = el.components.raycaster.objects.length;
       newEl.addEventListener('loaded', function () {
-        assert.equal(el.components.raycaster.objects.length, numObjects + 1);
-        done();
+        process.nextTick(function () {
```
Comment (dmarcos) [inline tests/components/raycaster.test.js]: Why do we need a nextTick here?

File: src/core/component.js (hunk 1 above)
Comment (ngokevin) [inline src/core/component.js]: OK. Seems like half our event names are hyphened and half are not (e.g., `enter-vr`). But I can keep it without here.

File: tests/components/raycaster.test.js (hunk 2 above)
Comment (ngokevin) [inline tests/components/raycaster.test.js]: Saw the test was failing. Maybe the raycaster handler was getting called after this handler.

Comment (ngokevin) [inline tests/components/raycaster.test.js]: Actually, could also change this tets to listen to child-attached.

**Comment by ngokevin** [comment]
re: event name hyphen/no space. Well, half the codebase uses hyphen, half no space. But I guess I'll keep it consistent within the module

**Comment by ngokevin** [comment]
updated. also added ID to componentremoved

File: tests/components/raycaster.test.js [hunk 3]
```diff
@@ -79,7 +79,7 @@ suite('raycaster', function () {
       var el = this.el;
       var newEl = this.targetEl = document.createElement('a-entity');
       var numObjects = el.components.raycaster.objects.length;
-      newEl.addEventListener('loaded', function () {
+      el.sceneEl.addEventListener('child-attached', function () {
```
Comment (dmarcos) [inline tests/components/raycaster.test.js]: We should remove dashes from event names for consistentcy

Comment (ngokevin) [inline tests/components/raycaster.test.js]: ok. will have to do that separately since a pretty large change.

**Comment by dmarcos** [comment]
Question. What's wrong with your `draw-line` having `movement-pattern` as dependency?

**Comment by ngokevin** [comment]
It might've been preemptive, but it's possible that component could work to draw any spline. Just wanted to decouple and make it reusable.

**Pull request merged**